src/Hamiltonians/HubbardRealSpace.jl

```diff
@@ -0,0 +1,351 @@
+        if i_a > i_b
+            # b is behind and needs to do a step
+            iter = iterate(occ_b, st_b)
+            isnothing(iter) && return acc * v
+            (n_b, i_b, _), st_b = iter
+        elseif i_a < i_b
+            # a is behind and needs to do a step
+            iter = iterate(occ_a, st_a)
+            isnothing(iter) && return acc * v
+            (n_a, i_a, _), st_a = iter
+        else
+            # a and b are at the same position
+            acc += n_a * n_b
+            # now both need to do a step
+            iter = iterate(occ_a, st_a)
+            isnothing(iter) && return acc * v
+            (n_a, i_a, _), st_a = iter
+            iter = iterate(occ_b, st_b)
+            isnothing(iter) && return acc * v
+            (n_b, i_b, _), st_b = iter
+        end
+    end
+end
+function real_space_interaction(b::BoseFS, f::FermiFS, v)
+    acc = 0
+    for (n, i) in occupied_modes(b)
+        acc += is_occupied(f, i) * n
+    end
+    return acc * v
+end
+function real_space_interaction(fs::CompositeFS, m)
+    _interactions(fs.components, m)
+end
+
+"""
+    _interaction_col(a, bs::Tuple, us::Tuple)
+
+Compute all interacitons in a column (below the diagonal). `a` is the address on the left-hand
+side of the interaction, and `bs` and `us` are the right-hand side addresses and interactions.
+"""
+@inline _interaction_col(a, ::Tuple{}, ::Tuple{}) = 0
+@inline function _interaction_col(a, (b, bs...), (u, us...))
+    return real_space_interaction(a, b, u) + _interaction_col(a, bs, us)
+end
+
+"""
+    _interaction(addresses, interaction_matrix)
+
+Compute all pairwise interactions in a tuple of `addresses`. The `interaction_matrix` sets the
+intraction strengths.
+
+The code is equivalent to the following.
+
+```julia
+acc = 0.0
+for (i, a) in enumerate(addresses)
+    acc += real_space_interaction(a, interaction_matrix[i, i])
+    for (j, b) in enumerate(addresses[i+1:end])
+        acc += real_space_interaction(a, b, interaction_matrix[i, j])
+    end
+end
+return acc
+```
+
+It is implemented recursively to ensure type stability.
+"""
+@inline _interactions(::Tuple{}, ::SMatrix{0,0}) = 0.0
+@inline function _interactions(
+    (a, as...)::NTuple{N,AbstractFockAddress}, m::SMatrix{N,N}
+) where {N}
+    # Split the matrix into the column we need now, and the rest.
+    (u, column...) = Tuple(m[:, 1])
+    # Type-stable way to subset SMatrix:
+    rest = SMatrix{N-1,N-1}(view(m, 2:N, 2:N))
+
+    # Get the self-interaction first.
+    self = real_space_interaction(a, u)
+    # Get the interactions for the rest of the row.
+    row = _interaction_col(a, as, column)
+    # Get the interaction for the rest of the rows.
+    return self + row + _interactions(as, rest)
+end
+
+###
+### HubbardRealSpace
+###
+"""
+    HubbardRealSpace(address; u=ones(C, C), t=ones(C), geometry=PeriodicBoundaries(M,))
+
+General Hubbard model in real space. Supports various different kinds of particle sizes and
+geometries.
```

Comment:
```suggestion
Hubbard model in real space. Supports single or multi-component Fock state 
addresses (with `C` components) and various (rectangular) lattice geometries.
```
It would be great to have a LaTeX formula for the general form of the lattice. Can you write one (starting with suggestion for `local_interaction`)?